File: advent of code/2021/java/day4.java
/**
 * Advent of Code 2021 - Day 4: Giant Squid
 * 3 December 2021 - 4 December 2021
 * 
 * part 1: 52 minutes
 * part 2: 35 minutes
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Scanner;

public class day4 {

    public static boolean checkForWin(int[][] board) {
        // checking rows
        for(int row = 0; row < 5; row++) {
            if(board[row][0] == board[row][1] &&
                board[row][0] == board[row][2] &&
                board[row][0] == board[row][3] &&
                board[row][0] == board[row][4]) {
                    return true;
            }

            if(board[0][row] == board[1][row] &&
                board[0][row] == board[2][row] &&
                board[0][row] == board[3][row] &&
                board[0][row] == board[4][row]) {
                    return true;
            }
        }

        return false;
    }

    public static void part1() throws FileNotFoundException {
        File input = new File("inputs/day4.txt");

        Scanner sc = new Scanner(input);

        String[] drawnStrings = sc.nextLine().split(",");
        ArrayList<Integer> drawn = new ArrayList<>();
        for(String d : drawnStrings)  {
            drawn.add(Integer.parseInt(d));
        } // all first line input

        ArrayList<int[][]> boards = new ArrayList<>();
        sc.nextLine();

        while(sc.hasNextLine()) {
            String currLine = sc.nextLine();
            if(currLine.isBlank()) {
                //System.out.println("BLANK");
            } else {
                int[][] currBoard = new int[5][5];
                for(int i = 0; i < 5; i++) {

                    String[] currArray = currLine.split("\\s\\s|\\s");

                        int counter = 0;
                        for(String line : currArray) {
                            if(!line.isEmpty()) {currBoard[i][counter] = Integer.parseInt(line);
                            counter++;}
                        }
                    
                    //System.out.println(currLine);
                    if(sc.hasNextLine()) {
                        currLine = sc.nextLine();
                    } else {
                        break;
                    }
                }
                boards.add(currBoard);
            }            
        }

        sc.close();

        // for(int[][] board : boards) {
        //     for(int i = 0; i < 5; i++) {
        //         for(int j = 0; j < 5; j++) {
        //             System.out.print(board[i][j] + " ");
        //         }
        //         System.out.println();
        //     }
        //     System.out.println();
        // }

        HashMap<int[][], LinkedList<Integer>> map = new HashMap<>();

        for(int[][] board : boards) {
            map.put(board, new LinkedList<Integer>());
        }

        // start putting in bingos

        int[][] winBoard = boards.get(0);

        do {
            int currNum = drawn.remove(0);
            for(int[][] board : boards) { // checking EACH BOARD
                
                for(int i = 0; i < 5; i++) {
                    for(int j = 0; j < 5; j++) {
                        if(board[i][j] == currNum) {
                            board[i][j] = 100;
                            map.get(board).add(currNum);
                            break;
                        }
                    }
                }

                if(checkForWin(board)) {
                    winBoard = board;
                }
                // boards.remove(currBoard);
                // boards.add(currBoard);
                // currBoard = boards.get(0);
            }

        } while(checkForWin(winBoard) == false);

        int unchecked = 0;

        for(int[] i : winBoard) {
            for(int j : i) {
                if(j != 100) {
                    unchecked += j;
                }
                System.out.print(j + " ");
            }
            System.out.println();
        }

        LinkedList<Integer> winList = map.get(winBoard);
        for(int val : winList) {
            System.out.print(val + " ");
        }
        System.out.println();

        System.out.println(unchecked);
        System.out.println(unchecked * winList.getLast());
    }
    public static void main(String[] args) throws FileNotFoundException {
        File input = new File("inputs/day4.txt");

        Scanner sc = new Scanner(input);

        String[] drawnStrings = sc.nextLine().split(",");
        ArrayList<Integer> drawn = new ArrayList<>();
        for(String d : drawnStrings)  {
            drawn.add(Integer.parseInt(d));
        } // all first line input

        ArrayList<int[][]> boards = new ArrayList<>();
        sc.nextLine();

        while(sc.hasNextLine()) {
            String currLine = sc.nextLine();
            if(currLine.isBlank()) {
                //System.out.println("BLANK");
            } else {
                int[][] currBoard = new int[5][5];
                for(int i = 0; i < 5; i++) {

                    String[] currArray = currLine.split("\\s\\s|\\s");

                        int counter = 0;
                        for(String line : currArray) {
                            if(!line.isEmpty()) {currBoard[i][counter] = Integer.parseInt(line);
                            counter++;}
                        }
                    
                    //System.out.println(currLine);
                    if(sc.hasNextLine()) {
                        currLine = sc.nextLine();
                    } else {
                        break;
                    }
                }
                boards.add(currBoard);
            }            
        }

        sc.close();

        HashMap<int[][], LinkedList<Integer>> map = new HashMap<>();

        for(int[][] board : boards) {
            map.put(board, new LinkedList<Integer>());
        }

        // start putting in bingos

        int[][] winBoard = boards.get(0);
        LinkedList<Integer> winList = map.get(winBoard);

        do {
            //System.out.println(map.size());

            int currNum = drawn.remove(0);
            for(int[][] board : boards) { // checking EACH BOARD
                
                if(map.containsKey(board)) {
                    for(int i = 0; i < 5; i++) {
                        for(int j = 0; j < 5; j++) {
                            if(board[i][j] == currNum) {
                                board[i][j] = 100;
                                map.get(board).add(currNum);

                                break;
                            }
                        }
                    }

                    if(checkForWin(board) && map.size() > 1) {
                        map.remove(board);
                    } else if(checkForWin(board) && map.size() == 1) {
                        System.out.println(map.size());
                        winList = map.get(board);
                        winBoard = board;
                        map.remove(board);
                    }
                }

                

                // boards.remove(currBoard);
                // boards.add(currBoard);
                // currBoard = boards.get(0);
            }
            //System.out.println(map.size());

            //System.out.println(map.size());

        } while(map.size() != 0);

        int unchecked = 0;

        for(int[] i : winBoard) {
            for(int j : i) {
                if(j != 100) {
                    unchecked += j;
                }
                System.out.print(j + " ");
            }
            System.out.println();
        }

        for(int val : winList) {
            System.out.print(val + " ");
        }
        System.out.println();

        System.out.println(unchecked);
        System.out.println(unchecked * winList.getLast());
    }
}

File: advent of code/2021/python/day4.py
# Advent of Code 2021 - Day 3: Binary Diagnostic
# 4 December 2021
# 12:12 PM - 12:19 PM (checkForWin)
# 12:40 PM - 1:10 PM
# 3:54 PM - 4:25

import re

def checkForWin(board):
    # checking rows
    for row in range(len(board)):
        if(board[row][0] == board[row][1] and
                board[row][0] == board[row][2] and
                board[row][0] == board[row][3] and
                board[row][0] == board[row][4]):
            return True
    
    for col in range(len(board[0])):
        if(board[0][col] == board[1][col] and 
                board[0][col] == board[2][col] and 
                board[0][col] == board[3][col] and 
                board[0][col] == board[4][col]):
            return True
    
    return False

def checkIfBoardHasNumber(board, number):
    for row in range(len(board)):
        for col in range(len(board[0])):
            if board[row][col] == number:
                board[row][col] = 100
                return True
    return False

def changeIfBoardHasNumber(board, number):
    for row in range(len(board)):
        for col in range(len(board[0])):
            if board[row][col] == number:
                board[row][col] = 100
                return board
    
# reading input
f = open('inputs/day4.txt', 'r')
lines = f.readlines()

numsCalled = lines.pop(0).rstrip().split(",")

for num in numsCalled: # changing all to int type
    num = int(num)

boards = []

while(len(lines) > 0):
    currLine = lines.pop(0).rstrip()

    currBoard = [[]*5 for x in range(5)] # makes an empty, two dimensional, 5x5 array

    if(currLine):
        for i in range(5):

            currCol = []

            colTemp = re.split( "\\s\\s|\\s", currLine)
            for val in colTemp:
                if(str(val)):
                    currCol.append(int(val))

            currBoard[i] = currCol
            if(len(lines) > 0): currLine = lines.pop(0)
        
        boards.append(currBoard)

# for board in boards:
#     for row in board:
#         for col in row:
#             print(col, end=" ")
        
#         print()
#     print()

# part 1

pickedNums = {}
winningBoard = None

while(winningBoard == None):
    currBingoNum = numsCalled.pop(0)
    #print(currBingoNum)

    for board in boards:

        print(board, currBingoNum)
        #print(checkIfBoardHasNumber(board, int(currBingoNum))) # somehow currBingoNum not an int?

        # ic ant get this code to fucking work

        # cannot hash immutable objects because its address changes when an item is added or removed

        if(checkIfBoardHasNumber(board, int(currBingoNum))):
            print("TRUE")


            currPicked = str(pickedNums[board])
            currPicked += str(currBingoNum, " ")
            pickedNums[board] = str(currPicked)
            board = changeIfBoardHasNumber(board, currBingoNum)
        
        # print(board)
        if(checkForWin(board)):
            winningBoard = board
            break

#print(checkIfBoardHasNumber([[30, 61, 64, 54, 80], [22, 47, 84, 16, 8], [83, 18, 65, 70, 11], [81, 23, 98, 26, 82], [45, 69, 6, 53, 68]], 70))
